### ISSUE TITLE
PDF/UA accessebility. Table bug 

### ISSUE THREAD

**Issue by bandirom** [Kozea/WeasyPrint#1784]
Description issue link

#1088 

[https://github.com/Kozea/WeasyPrint/issues/1088#issuecomment-1340867520](url)

Thank you for your dedication!

**Comment by bandirom** [comment]
Hello!) do you have any updates?

**Comment by liZe** [comment]
> Hello!) do you have any updates?

Hi!

There’s nothing new on this topic. There’s definitely a bug, but we didn’t take the time to solve it yet.

If anyone is interested, the code that needs to be changed is probably here:

https://github.com/Kozea/WeasyPrint/blob/79e84dadff11c60a278c031881f65e5f0ff41b92/weasyprint/pdf/pdfua.py#L90

Just as there’s a special case for `LI` elements that requires a special structure, there’s probably a special case for table elements that need their own specific structure.

**Comment by bandirom** [comment]
Some updates:
I used [https://www.pdfa.org/](https://www.pdfa.org/) and received next info about PDF/UA-1

![image](https://user-images.githubusercontent.com/46032838/216588080-e3e1a8f3-791c-4a03-8d95-0e27572f70da.png)

**Comment by ckpearson** [comment]
Any progress on this? Being able to produce compliant pdf/ua output is pretty crucial for our use case (generating accessible media for blind & low vision users), I'm specifying headers & scope attributes for my tables, but it looks like the `Headers` attribute isn't making it in to the document.

**Comment by liZe** [comment]
> Any progress on this? Being able to produce compliant pdf/ua output is pretty crucial for our use case

Nothing new since https://github.com/Kozea/WeasyPrint/issues/1784#issuecomment-1380199206.

Debugging PDF/UA is really complicated for us right now, mainly because we don’t know the specification very well, and because we don’t have the right tools to check the result. Help and hints are welcome for these two points. If you know Python, you can also try to read the related code and propose a fix.

This is the kind of topics we’d **love** to spend more time on, but it requires long periods of continuous time to dive into the problem seriously. Unfortunately, that’s not something we can usually do on our free time.

**Comment by ckpearson** [comment]
Understandable, I'm pretty new to it myself, I'm using veraPdf (https://verapdf.org) for the validation piece. I'd be open to opening a PR if I can get my head around how it's structured and hangs together, I'm not super familiar with the spec or python myself, and would need to understand how weasy print tracks element types as it builds the tree and decorates with metadata.

**Comment by ckpearson** [comment]
As a note: https://github.com/veraPDF/veraPDF-rest?tab=readme-ov-file it looks like they now have a docker image you can spin up and submit PDFs via a POST request to receive validation reports back from.

**Comment by liZe** [comment]
Thanks for the hint. We already use veraPDF and it works well, but it’s sometimes hard to understand why (and where) things are wrong in the PDF. It’s also really hard for us to see the structure of tags in the PDFs we generate.

**Comment by ckpearson** [comment]
No worries, I've just found PDFbox's debugger tool which might be of some use:

![Screenshot 2024-09-10 at 21 49 46](https://github.com/user-attachments/assets/2a9cf937-af99-4c6c-a2f0-afc99a7bb3ec)
![Screenshot 2024-09-10 at 21 50 07](https://github.com/user-attachments/assets/7a92e050-f9c7-41b8-8472-8e2d06304991)

Would the weasyprint project be open to possible sponsorship / commercial collaboration around helping make improvements down the line? Can't make any guarantees but our use-case is for a rich PDF production capability across our entire estate, with a key focus being accessibility and weasyprint seems to get us the closest of the OSS tools currently available.

**Comment by ckpearson** [comment]
Should say this is the Apache PDFBox (https://pdfbox.apache.org/download.html) standalone debugger tool.

**Comment by liZe** [comment]
> Would the weasyprint project be open to possible sponsorship / commercial collaboration around helping make improvements down the line? Can't make any guarantees but our use-case is for a rich PDF production capability across our entire estate, with a key focus being accessibility and weasyprint seems to get us the closest of the OSS tools currently available.

Yes! Please [contact CourtBouillon (us) by mail](https://www.courtbouillon.org/index/#support), we’ll be happy to talk about this topic with you.

> Should say this is the Apache PDFBox (https://pdfbox.apache.org/download.html) standalone debugger tool.

Looks interesting, thanks for sharing!

**Comment by ckpearson** [comment]
I've also found this https://github.com/itext/i7j-rups/releases/tag/7.2.5 which the PDF Association (https://github.com/itext/i7j-rups/releases/tag/7.2.5) apparently uses internally. You can also get hold of the ISO specification document for pdf/ua-1 from them for nothing too!

I've had a soft approval for the sponsorship piece also, so will definitely be in touch.

**Comment by liZe** [comment]
If you’re interested in this issue, it’s time to test #2471! Feedback will be highly appreciated, even to say that it just works. 🙏
